app/boot_control/cboot.py

```diff
@@ -0,0 +1,525 @@
+
+    # nvbootctrl
+    @staticmethod
+    def _nvbootctrl(arg: str, *, call_only=True, raise_exception=True) -> Optional[str]:
+        """
+        Raises:
+            NvbootCtrlError if raise_exception is True.
+        """
+        # NOTE: target is always set to rootfs
+        _cmd = f"nvbootctrl -t rootfs {arg}"
+        try:
+            if call_only:
+                subprocess_call(_cmd, raise_exception=raise_exception)
+                return
+            else:
+                return subprocess_check_output(_cmd, raise_exception=raise_exception)
+        except CalledProcessError as e:
+            raise NvbootctrlError from e
+
+    @classmethod
+    def check_rootdev(cls, dev: str) -> bool:
+        """
+        check whether the givin dev is legal root dev or not
+
+        NOTE: expect using UUID method to assign rootfs!
+        """
+        pa = re.compile(r"\broot=(?P<rdev>[\w=-]*)\b")
+        try:
+            if ma := pa.search(subprocess_check_output("cat /proc/cmdline")):
+                res = ma.group("rdev")
+            else:
+                raise ValueError(f"failed to find specific {dev} in cmdline")
+        except (CalledProcessError, ValueError) as e:
+            raise NvbootctrlError(
+                "rootfs detect failed or rootfs is not specified by PARTUUID in kernel cmdline"
+            ) from e
+
+        uuid = res.split("=")[-1]
+
+        return Path(CMDHelperFuncs.get_dev_by_partuuid(uuid)).resolve(
+            strict=True
+        ) == Path(dev).resolve(strict=True)
+
+    @classmethod
+    def get_current_slot(cls) -> str:
+        if slot := cls._nvbootctrl("get-current-slot", call_only=False):
+            return slot
+        else:
+            raise NvbootctrlError
+
+    @classmethod
+    def mark_boot_successful(cls, slot: str):
+        cls._nvbootctrl(f"mark-boot-successful {slot}")
+
+    @classmethod
+    def set_active_boot_slot(cls, slot: str):
+        cls._nvbootctrl(f"set-active-boot-slot {slot}")
+
+    @classmethod
+    def set_slot_as_unbootable(cls, slot: str):
+        cls._nvbootctrl(f"set-slot-as-unbootable {slot}")
+
+    @classmethod
+    def is_slot_bootable(cls, slot: str) -> bool:
+        try:
+            cls._nvbootctrl(f"is-slot-bootable {slot}")
+            return True
+        except NvbootctrlError:
+            return False
+
+    @classmethod
+    def is_slot_marked_successful(cls, slot: str) -> bool:
+        try:
+            cls._nvbootctrl(f"is-slot-marked-successful {slot}")
+            return True
+        except NvbootctrlError:
+            return False
+
+
+class _CBootControl:
+    KERNEL: str = "/boot/Image"
+    KERNEL_SIG: str = "/boot/Image.sig"
+    INITRD: str = "/boot/initrd"
+    INITRD_IMG_LINK: str = "/boot/initrd.img"
+    FDT: str = "/boot/tegra194-rqx-580.dtb"
+    FDT_HDR40: str = "/boot/tegra194-rqx-580-hdr40.dtbo"
+    EXTRA_CMDLINE: str = (
+        "console=ttyTCU0,115200n8 console=tty0 fbcon=map:0 net.ifnames=0"
+    )
```

Comment:
Delete these lines, since they are not in use. This definition is scary because extlinux.conf is not parsed.

Reply:
updated in 1e6c6dd